// asn1.js/lib/asn1/api.js
'use strict';

const encoders = require('./encoders');
const inherits = require('inherits');

const api = exports;

api.define = function define() {
  return new Entity();
};

function Entity() {
  this.body = function() {
    this.seq().obj(
      this.key('version').int(),
      this.key('privateKey').octstr(),
      this.key('parameters').explicit(0).objid()
        .optional(),
      this.key('publicKey').explicit(1).bitstr()
        .optional()
    );
  }

  this.encoders = {};
}

Entity.prototype._createNamed = function createNamed(Base) {

  function Generated(entity) {
    this._initNamed(entity);
  }
  inherits(Generated, Base);
  Generated.prototype._initNamed = function _initNamed(entity) {
    Base.call(this, entity, 'ECPrivateKey');
  };

  return new Generated(this);
};

Entity.prototype.encode = function encode(data) {
    console.log("Entity.prototype.encode");
    console.log(this.encoders.pem);

    if (!this.encoders.pem)
        this.encoders.pem = this._createNamed(encoders.pem);


    console.log(this.encoders.pem);
    return this.encoders.pem.encode(data);
};
