// web-push/src/web-push-lib.js
'use strict';

const url = require('url');
const https = require('https');
const asn1 = require('asn1.js');
const jws = require('jws');

const WebPushError = require('./web-push-error.js');
const vapidHelper = require('./vapid-helper.js');
const encryptionHelper = require('./encryption-helper.js');
const webPushConstants = require('./web-push-constants.js');
const urlBase64Helper = require('./urlsafe-base64-helper');

// Default TTL is four weeks.
const DEFAULT_TTL = 2419200;

let gcmAPIKey = '';
let vapidDetails;

function WebPushLib() {

}

/**
 * When sending messages to a GCM endpoint you need to set the GCM API key
 * by either calling setGMAPIKey() or passing in the API key as an option
 * to sendNotification().
 * @param  {string} apiKey The API key to send with the GCM request.
 */
WebPushLib.prototype.setGCMAPIKey = function(apiKey) {
  if (apiKey === null) {
    gcmAPIKey = null;
    return;
  }

  if (typeof apiKey === 'undefined'
  || typeof apiKey !== 'string'
  || apiKey.length === 0) {
    throw new Error('The GCM API Key should be a non-empty string or null.');
  }

  gcmAPIKey = apiKey;
};

/**
 * When making requests where you want to define VAPID details, call this
 * method before sendNotification() or pass in the details and options to
 * sendNotification.
 * @param  {string} subject    This must be either a URL or a 'mailto:'
 * address. For example: 'https://my-site.com/contact' or
 * 'mailto: contact@my-site.com'
 * @param  {string} publicKey  The public VAPID key, a URL safe, base64 encoded string
 * @param  {string} privateKey The private VAPID key, a URL safe, base64 encoded string.
 */
WebPushLib.prototype.setVapidDetails = function(subject, publicKey, privateKey) {
    if (arguments.length === 1 && arguments[0] === null) {
      vapidDetails = null;
      return;
    }

    vapidHelper.validateSubject(subject);
    vapidHelper.validatePublicKey(publicKey);
    vapidHelper.validatePrivateKey(privateKey);

    vapidDetails = {
      subject: subject,
      publicKey: publicKey,
      privateKey: privateKey
    };
  };

function getFutureExpirationTimestamp(numSeconds) {
  const futureExp = new Date();
  futureExp.setSeconds(futureExp.getSeconds() + numSeconds);
  return Math.floor(futureExp.getTime() / 1000);
}

const ECPrivateKeyASN = asn1.define('ECPrivateKey', function() {
  this.seq().obj(
    this.key('version').int(),
    this.key('privateKey').octstr(),
    this.key('parameters').explicit(0).objid()
      .optional(),
    this.key('publicKey').explicit(1).bitstr()
      .optional()
  );
});

const DEFAULT_EXPIRATION_SECONDS = 12 * 60 * 60;

const supportedContentEncodings = {
        AES_GCM: 'aesgcm',
        AES_128_GCM: 'aes128gcm'
      };
WebPushLib.prototype.generateRequestDetails = function(subscription, payload, options) {
    let contentEncoding = supportedContentEncodings.AES_128_GCM;

    const requestDetails = { method: 'POST', headers: { TTL: DEFAULT_TTL } };


    requestDetails.headers['Content-Length'] = 0;

    const parsedUrl = url.parse(subscription.endpoint);
    const audience = parsedUrl.protocol + '//' + parsedUrl.host;

    console.log(vapidDetails);
//    const vapidHeaders = vapidHelper.getVapidHeaders(
//        audience,
//        vapidDetails.subject,
//        vapidDetails.publicKey,
//        vapidDetails.privateKey,
//        contentEncoding);

    const publicKey = Buffer.from(vapidDetails.publicKey, 'base64url');
    const privateKey = Buffer.from(vapidDetails.privateKey, 'base64url');

    const expiration = getFutureExpirationTimestamp(DEFAULT_EXPIRATION_SECONDS);

    const header = {
        typ: 'JWT',
        alg: 'ES256'
    };

    const jwtPayload = {
        aud: audience,
        exp: expiration,
        sub: vapidDetails.subject
    };
    console.log(jwtPayload);
    console.log(privateKey);
    const pemKey = //toPEM(privateKey);
        ECPrivateKeyASN.encode({
            version: 1,
            privateKey: privateKey,
            parameters: [1, 2, 840, 10045, 3, 1, 7] // prime256v1
          }, 'pem', {
            label: 'EC PRIVATE KEY'
          });

    console.log(pemKey);

    const jwt = jws.sign({
        header: header,
        payload: jwtPayload,
        privateKey: pemKey
    });
    console.log(jwt);

    const vapidHeaders = {
      Authorization: 'vapid t=' + jwt + ', k=' + vapidDetails.publicKey
    };

    console.log(vapidHeaders);
    requestDetails.headers.Authorization = vapidHeaders.Authorization;

    requestDetails.headers.Urgency = webPushConstants.supportedUrgency.NORMAL;
    requestDetails.endpoint = subscription.endpoint;

    return requestDetails;
  };

/**
 * To send a push notification call this method with a subscription, optional
 * payload and any options.
 * @param  {PushSubscription} subscription The PushSubscription you wish to
 * send the notification to.
 * @param  {string|Buffer} [payload]       The payload you wish to send to the
 * the user.
 * @param  {Object} [options]              Options for the GCM API key and
 * vapid keys can be passed in if they are unique for each notification you
 * wish to send.
 * @return {Promise}                       This method returns a Promise which
 * resolves if the sending of the notification was successful, otherwise it
 * rejects.
 */
WebPushLib.prototype.sendNotification = function(subscription, payload, options) {
    let requestDetails;
    try {
      requestDetails = this.generateRequestDetails(subscription, payload, options);
    } catch (err) {
      return Promise.reject(err);
    }

    return new Promise(function(resolve, reject) {
      const httpsOptions = {};
      const urlParts = url.parse(requestDetails.endpoint);
      httpsOptions.hostname = urlParts.hostname;
      httpsOptions.port = urlParts.port;
      httpsOptions.path = urlParts.path;

      httpsOptions.headers = requestDetails.headers;
      httpsOptions.method = requestDetails.method;

      if (requestDetails.timeout) {
        httpsOptions.timeout = requestDetails.timeout;
      }

      if (requestDetails.agent) {
        httpsOptions.agent = requestDetails.agent;
      }

      if (requestDetails.proxy) {
        const { HttpsProxyAgent } = require('https-proxy-agent'); // eslint-disable-line global-require
        httpsOptions.agent = new HttpsProxyAgent(requestDetails.proxy);
      }

      const pushRequest = https.request(httpsOptions, function(pushResponse) {
        let responseText = '';

        pushResponse.on('data', function(chunk) {
          responseText += chunk;
        });

        pushResponse.on('end', function() {
          if (pushResponse.statusCode < 200 || pushResponse.statusCode > 299) {
            reject(new WebPushError(
              'Received unexpected response code',
              pushResponse.statusCode,
              pushResponse.headers,
              responseText,
              requestDetails.endpoint
            ));
          } else {
            resolve({
              statusCode: pushResponse.statusCode,
              body: responseText,
              headers: pushResponse.headers
            });
          }
        });
      });

      if (requestDetails.timeout) {
        pushRequest.on('timeout', function() {
          pushRequest.destroy(new Error('Socket timeout'));
        });
      }

      pushRequest.on('error', function(e) {
        reject(e);
      });

      if (requestDetails.body) {
        pushRequest.write(requestDetails.body);
      }

      pushRequest.end();
    });
  };

module.exports = WebPushLib;
